### PR TITLE
Enable fraction to limit dataset by image counts

### PR DESCRIPTION
**Pull request by fcakyon**
Training fixed-size subsets currently requires separate dataset copies, and Platform NDJSON downloads every image before fraction is applied.

model.train(data="ul://...", fraction=1000) now caps train to 1,000 images. fraction=[1000, 100] caps train and val.

- filters count-based NDJSON records before download
- leaves float ratio behavior unchanged

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Extended `fraction` to accept positive image counts and `[train, val]` split limits, including pre-download filtering for NDJSON datasets while preserving float-ratio behavior.

### 📊 Key Changes

- Training now accepts values such as `fraction=1000` for a fixed image count or `fraction=[1000, 100]` for separate train and validation limits.
- Added split-aware handling across YOLO, RTDETR, classification, validation, and INT8 calibration dataloaders, with integer `1` treated as one image and float `1.0` as the full split.
- NDJSON conversion filters sorted train and validation records using evenly spaced selection before downloading, keeps test records complete, and includes the selection in conversion caching.
- Configuration validation now accepts positive integer counts and two-value train/validation lists for `fraction`; related training and export documentation was updated.
- Training logs now report the selected train and validation image counts and configured `fraction`.

### 🎯 Purpose & Impact

- Users can run fixed-size training, validation, or calibration subsets without creating separate dataset copies.
- Platform and other NDJSON-backed datasets download only the selected train and validation records, while float ratio selection remains supported and test data remains full.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fcakyon, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The train/val cap implementation is directionally correct, but the current diff drops test records, can infer an incorrect class count from capped records, applies the validation cap to test splits, and introduces a callable compatibility regression. These should be fixed before merge.

<details><summary>💬 Posted 4 inline comments</summary>

- 💡 **MEDIUM** `ultralytics/data/converter.py:989` This rebuilds `image_records` exclusively from `train` and `val`, so valid `test` records are discarded whenever `fraction` is an integer or list. The converter explicitly accepts the `test` split, and `model.val(..., split="test")` will consequently fail with missing test data or silently lose that split. Preserve uncapped splits while limiting only the requested train/val records.
- 💡 **MEDIUM** `ultralytics/data/converter.py:995` The count filter runs before the existing class-count inference. For an NDJSON manifest without `class_names`—a supported path—if the highest class ID occurs outside the selected prefix, the generated YAML infers an undersized `nc` (for example, `nc=2` instead of `nc=3`), producing a model with incomplete class metadata. Infer the dataset class schema from the complete manifest before applying the image cap.
- 💡 **MEDIUM** `ultralytics/data/utils.py:518` `split != "train"` maps the validation count to every other split, including `test`. Thus `fraction=[1000, 100]` silently truncates test evaluation to 100 images for local YAML/classification datasets, although the documented pair only specifies train and val counts. Handle `test` separately so it retains its prior uncapped behavior unless explicitly supported.
- 📝 **LOW** `ultralytics/data/utils.py:521` This changes a non-underscored utility from accepting one argument to requiring `fraction`, so existing callers of `convert_ndjson_to_yolo_if_needed(data)` now raise `TypeError`. Keep the default conversion behavior backward-compatible by defaulting `fraction` to `1.0`.
</details>

File: ultralytics/data/converter.py [hunk 1]
```diff
@@ -984,6 +986,15 @@ def read_records():
                 f"For best results, manually assign validation images in Platform dataset page."
             )
 
+    if isinstance(fraction, (int, list)):
```
Comment (UltralyticsAssistant) [inline ultralytics/data/converter.py]: 💡 **MEDIUM**: This rebuilds `image_records` exclusively from `train` and `val`, so valid `test` records are discarded whenever `fraction` is an integer or list. The converter explicitly accepts the `test` split, and `model.val(..., split="test")` will consequently fail with missing test data or silently lose that split. Preserve uncapped splits while limiting only the requested train/val records.

File: ultralytics/data/converter.py [hunk 2]
```diff
@@ -984,6 +986,15 @@ def read_records():
...
+        counts = fraction if isinstance(fraction, list) else (fraction, None)
+        image_records = [
+            record
+            for split, count in zip(("train", "val"), counts)
+            for record in [r for r in image_records if r["split"] == split][:count]
+        ]
```
Comment (UltralyticsAssistant) [inline ultralytics/data/converter.py]: 💡 **MEDIUM**: The count filter runs before the existing class-count inference. For an NDJSON manifest without `class_names`—a supported path—if the highest class ID occurs outside the selected prefix, the generated YAML infers an undersized `nc` (for example, `nc=2` instead of `nc=3`), producing a model with incomplete class metadata. Infer the dataset class schema from the complete manifest before applying the image cap.

File: ultralytics/data/utils.py [hunk 3]
```diff
@@ -513,7 +513,12 @@ def find_dataset_yaml(path: Path) -> Path:
     return files[0]
 
 
-def convert_ndjson_to_yolo_if_needed(data: str | Path) -> str | Path:
+def get_split_fraction(fraction: float | list[int], split: str) -> float | int:
+    """Return the scalar dataset fraction for a train or validation split."""
+    return fraction[split != "train"] if isinstance(fraction, list) else fraction if split == "train" else 1.0
```
Comment (UltralyticsAssistant) [inline ultralytics/data/utils.py]: 💡 **MEDIUM**: `split != "train"` maps the validation count to every other split, including `test`. Thus `fraction=[1000, 100]` silently truncates test evaluation to 100 images for local YAML/classification datasets, although the documented pair only specifies train and val counts. Handle `test` separately so it retains its prior uncapped behavior unless explicitly supported.

File: ultralytics/data/utils.py [hunk 4]
```diff
@@ -513,7 +513,12 @@ def find_dataset_yaml(path: Path) -> Path:
...
+
+
+def convert_ndjson_to_yolo_if_needed(data: str | Path, fraction: float | list[int]) -> str | Path:
```
Comment (UltralyticsAssistant) [inline ultralytics/data/utils.py]: 📝 **LOW**: This changes a non-underscored utility from accepting one argument to requiring `fraction`, so existing callers of `convert_ndjson_to_yolo_if_needed(data)` now raise `TypeError`. Keep the default conversion behavior backward-compatible by defaulting `fraction` to `1.0`.

**Suggested change:**
```suggestion
def convert_ndjson_to_yolo_if_needed(data: str | Path, fraction: float | int | list[int] = 1.0) -> str | Path:
```

File: ultralytics/data/converter.py [hunk 5]
```diff
@@ -984,6 +986,15 @@ def read_records():
                 f"For best results, manually assign validation images in Platform dataset page."
             )
 
+    if isinstance(fraction, (int, list)):
+        counts = fraction if isinstance(fraction, list) else (fraction, None)
+        image_records = [
```
Comment (UltralyticsAssistant) [inline ultralytics/data/converter.py]: 💡 **MEDIUM**: An NDJSON manifest with a `test` split is silently stripped whenever `fraction` is an integer or list because this rebuilds `image_records` from only `train` and `val`. A subsequent `model.val(split="test", fraction=...)` then fails with a missing split instead of evaluating the test set. Slice only `train`/`val` records while preserving other splits, especially `test`.

File: ultralytics/data/utils.py (hunk 3 above)
Comment (UltralyticsAssistant) [inline ultralytics/data/utils.py]: 💡 **MEDIUM**: `split != "train"` indexes the validation count for every non-training split. Thus `fraction=[1000, 100]` also caps `split="test"` to 100 images, although the pair is defined as `[train, val]`. Return the second count only for `split == "val"` and leave test uncapped.

File: ultralytics/models/rtdetr/train.py [hunk 6]
```diff
@@ -82,7 +83,7 @@ def build_dataset(self, img_path: str, mode: str = "val", batch: int | None = No
             prefix=colorstr(f"{mode}: "),
             classes=self.args.classes,
             data=self.data,
-            fraction=self.args.fraction if mode == "train" else 1.0,
+            fraction=get_split_fraction(self.args.fraction, mode),
```
Comment (UltralyticsAssistant) [inline ultralytics/models/rtdetr/train.py]: 💡 **MEDIUM**: This propagates the split-specific fraction only through `RTDETRTrainer`. Standalone `RTDETRValidator.build_dataset()` still constructs `RTDETRDataset` without a `fraction`, so `model.val(fraction=[1000, 100])` evaluates the full validation set instead of applying the val cap. Pass the corresponding split fraction in the RT-DETR validator as well.

File: ultralytics/data/utils.py (hunk 4 above)
Comment (UltralyticsAssistant) [inline ultralytics/data/utils.py]: 📝 **LOW**: This changes a documented utility's existing call contract: callers that use `convert_ndjson_to_yolo_if_needed(data)` now get a `TypeError` before the function can inspect the input. Preserve the old one-argument behavior with the existing default of `1.0`, while callers needing caps pass the new argument explicitly.

**Suggested change:**
```suggestion
def convert_ndjson_to_yolo_if_needed(data: str | Path, fraction: float | int | list[int] = 1.0) -> str | Path:
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all 16 changed files and traced fraction handling through configuration, NDJSON conversion, dataset construction, validation, export calibration, and RT-DETR. The train/val count-cap path is generally wired correctly, but test splits are mishandled, standalone RT-DETR validation misses the cap, and a documented utility has an avoidable API regression.

<details><summary>💬 Posted 4 inline comments</summary>

- 💡 **MEDIUM** `ultralytics/data/converter.py:991` An NDJSON manifest with a `test` split is silently stripped whenever `fraction` is an integer or list because this rebuilds `image_records` from only `train` and `val`. A subsequent `model.val(split="test", fraction=...)` then fails with a missing split instead of evaluating the test set. Slice only `train`/`val` records while preserving other splits, especially `test`.
- 💡 **MEDIUM** `ultralytics/data/utils.py:518` `split != "train"` indexes the validation count for every non-training split. Thus `fraction=[1000, 100]` also caps `split="test"` to 100 images, although the pair is defined as `[train, val]`. Return the second count only for `split == "val"` and leave test uncapped.
- 💡 **MEDIUM** `ultralytics/models/rtdetr/train.py:86` This propagates the split-specific fraction only through `RTDETRTrainer`. Standalone `RTDETRValidator.build_dataset()` still constructs `RTDETRDataset` without a `fraction`, so `model.val(fraction=[1000, 100])` evaluates the full validation set instead of applying the val cap. Pass the corresponding split fraction in the RT-DETR validator as well.
- 📝 **LOW** `ultralytics/data/utils.py:521` This changes a documented utility's existing call contract: callers that use `convert_ndjson_to_yolo_if_needed(data)` now get a `TypeError` before the function can inspect the input. Preserve the old one-argument behavior with the existing default of `1.0`, while callers needing caps pass the new argument explicitly.
</details>

File: ultralytics/data/converter.py (hunk 1 above)
Comment (fcakyon) [reply ultralytics/data/converter.py]: fixed in [a88d5cfe1](https://github.com/ultralytics/ultralytics/commit/a88d5cfe14ed04809621d1df45dc4e491fb5db63): test records stay uncapped

File: ultralytics/data/converter.py (hunk 2 above)
Comment (fcakyon) [reply ultralytics/data/converter.py]: fixed in [a88d5cfe1](https://github.com/ultralytics/ultralytics/commit/a88d5cfe14ed04809621d1df45dc4e491fb5db63): class metadata is inferred before capping

File: ultralytics/data/utils.py (hunk 3 above)
Comment (fcakyon) [reply ultralytics/data/utils.py]: fixed in [a88d5cfe1](https://github.com/ultralytics/ultralytics/commit/a88d5cfe14ed04809621d1df45dc4e491fb5db63): list counts apply only to train and val

File: ultralytics/data/utils.py (hunk 4 above)
Comment (fcakyon) [reply ultralytics/data/utils.py]: fixed in [a88d5cfe1](https://github.com/ultralytics/ultralytics/commit/a88d5cfe14ed04809621d1df45dc4e491fb5db63): restored the 1.0 default

File: ultralytics/data/converter.py (hunk 5 above)
Comment (fcakyon) [reply ultralytics/data/converter.py]: fixed in [a88d5cfe1](https://github.com/ultralytics/ultralytics/commit/a88d5cfe14ed04809621d1df45dc4e491fb5db63): test records stay uncapped

File: ultralytics/data/utils.py (hunk 3 above)
Comment (fcakyon) [reply ultralytics/data/utils.py]: fixed in [a88d5cfe1](https://github.com/ultralytics/ultralytics/commit/a88d5cfe14ed04809621d1df45dc4e491fb5db63): list counts apply only to train and val

File: ultralytics/models/rtdetr/train.py (hunk 6 above)
Comment (fcakyon) [reply ultralytics/models/rtdetr/train.py]: fixed in [db200d214](https://github.com/ultralytics/ultralytics/commit/db200d214fb685b3fc3d40bb1db92384876178dc): standalone RT-DETR validation now receives the split cap

File: ultralytics/data/utils.py (hunk 4 above)
Comment (fcakyon) [reply ultralytics/data/utils.py]: fixed in [a88d5cfe1](https://github.com/ultralytics/ultralytics/commit/a88d5cfe14ed04809621d1df45dc4e491fb5db63): restored the 1.0 default

**Review by glenn-jocher — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
🎉 This PR has been merged! Thank you @fcakyon, with contributions from @glenn-jocher, for [PR #25951](https://github.com/ultralytics/ultralytics/pull/25951).

> “Alone we can do so little; together we can do so much.” — Helen Keller

This collaboration makes dataset subset selection more flexible, supporting fixed image counts and split-specific limits across training, validation, calibration, and NDJSON pre-download workflows while preserving float-ratio behavior and complete test data.
